### PR TITLE
[SPARK-12975] [SQL] Throwing Exception when Bucketing Columns are part of Partitioning Columns

### PR DESCRIPTION
**Pull request by gatorsmile**
When users are using `partitionBy` and `bucketBy` at the same time, some bucketing columns might be part of partitioning columns. For example, 

```
        df.write
          .format(source)
          .partitionBy("i")
          .bucketBy(8, "i", "k")
          .saveAsTable("bucketed_table")
```

However, in the above case, adding column `i` into `bucketBy` is useless. It is just wasting extra CPU when reading or writing bucket tables. Thus, like Hive, we can issue an exception and let users do the change. 

Also added a test case for checking if the information of `sortBy` and `bucketBy` columns are correctly saved in the metastore table. 

Could you check if my understanding is correct? @cloud-fan @rxin @marmbrus Thanks!

### CODE REVIEW

**Comment by rxin** [comment]
Does Hive write them out?

**Comment by cloud-fan** [comment]
Will hive throw exception for this?

**Comment by gatorsmile** [comment]
I am not a Hive expert. I just did a try in Hive 1.2.1:
```
hive> CREATE TABLE user_info_bucketed(user_id BIGINT, firstname STRING, lastname STRING)
    > PARTITIONED BY(ds STRING)
    > CLUSTERED BY(ds, user_id) INTO 256 BUCKETS;
FAILED: SemanticException [Error 10002]: Invalid column reference
```

It sounds like Hive does not allow users use Partitioning columns in bucketing key. I think this is not an issue in Hive. However, this is not prohibited in our Spark SQL.  @rxin @cloud-fan

**Comment by cloud-fan** [comment]
I think we should follow hive for this case, i.e. throw exception.

**Comment by gatorsmile** [comment]
Ok, let users change it. Will do the change.

**Comment by gatorsmile** [comment]
Also updated the description of PR. Code is ready for review. : ) @cloud-fan Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -240,6 +241,15 @@ final class DataFrameWriter private[sql](df: DataFrame) {
       n <- numBuckets
     } yield {
       require(n > 0 && n < 100000, "Bucket number must be greater than 0 and less than 100000.")
+
+      // partitionBy columns cannot be used in blockedBy
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: typo: blockedBy.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala [hunk 2]
```diff
@@ -776,6 +776,59 @@ class MetastoreDataSourcesSuite extends QueryTest with SQLTestUtils with TestHiv
...
+
+    withTable(tableName) {
+      df.write
+        .format("parquet")
+        .bucketBy(8, "d", "b")
+        .sortBy("c")
+        .saveAsTable(tableName)
+      invalidateTable(tableName)
+      val metastoreTable = catalog.client.getTable("default", tableName)
+      val expectedBlockByColumns = StructType(df.schema("d") :: df.schema("b") :: Nil)
```
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala]: typo too: BlockBy.
There are many same typos below.

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala]: : ) Thank you for catching it. Will do a change. BlockBy is used in a tool I developed last year.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 3]
```diff
@@ -37,7 +38,7 @@ import org.apache.spark.sql.sources.HadoopFsRelation
  * @since 1.4.0
  */
 @Experimental
-final class DataFrameWriter private[sql](df: DataFrame) {
+final class DataFrameWriter private[sql](df: DataFrame) extends Logging  {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: no need to do this?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala [hunk 4]
```diff
@@ -776,6 +776,59 @@ class MetastoreDataSourcesSuite extends QueryTest with SQLTestUtils with TestHiv
     }
   }
 
+  test("Saving information for sortBy and bucketBy columns") {
+    val df = (1 to 10).map(i => (i, i + 1, s"str$i", s"str${i + 1}")).toDF("a", "b", "c", "d")
+    val tableName = s"partitionInfo_${System.currentTimeMillis()}"
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala]: nit: use `bucketingInfo` to be more accuracy.

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/BucketedWriteSuite.scala [hunk 5]
```diff
@@ -158,6 +161,21 @@ class BucketedWriteSuite extends QueryTest with SQLTestUtils with TestHiveSingle
     }
   }
 
+  test("write bucketed data with the overlapping bucketBy and partitionBy columns") {
+    intercept[AnalysisException](df.write
+      .partitionBy("i")
+      .bucketBy(8, "i", "k")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/sources/BucketedWriteSuite.scala]: is it better to use `i, j` and `j ,k` for "overlapping"?

**Comment by cloud-fan** [comment]
LGTM exception some minor comments, thanks for working on it!

**Comment by gatorsmile** [comment]
Thank you for your review! So glad I can provide a help in this great feature! Table bucketing is very useful in real-world scenario.

**Comment by cloud-fan** [comment]
LGTM

**Comment by marmbrus** [comment]
Thanks, merging to master.

**Pull request closed without merging**